parsers/CA_NS.py

```diff
@@ -132,17 +132,25 @@ def fetch_exchange(
     soup = BeautifulSoup(session.get(EXCHANGE_URL).text, "html.parser")
 
     # Extract the timestamp from the table header.
-    timestamp = datetime.strptime(
-        soup.find(string="Current System Conditions").find_next("td").em.i.string,
-        "%d-%b-%y %H:%M:%S",
-    ).replace(tzinfo=ZoneInfo("America/Halifax"))
+    try:
+        timestamp = datetime.strptime(
+            soup.find(string="Current System Conditions").find_next("td").em.i.string,
+            "%d-%b-%y %H:%M:%S",
+        ).replace(tzinfo=ZoneInfo("America/Halifax"))
+    except (AttributeError, TypeError, ValueError):
+        logger.error("unable to extract timestamp (error: {error})")
+        return []
 
     # Choose the appropriate exchange figure for the requested zone pair.
-    exchange = (
-        -float(soup.find(string="NS Export ").find_next("td").string)
-        if sorted_zone_keys == ZoneKey("CA-NB->CA-NS")
-        else float(soup.find(string="Maritime Link Import ").find_next("td").string)
-    )
+    try:
+        exchange = (
+            -float(soup.find(string="NS Export ").find_next("td").string)
+            if sorted_zone_keys == ZoneKey("CA-NB->CA-NS")
+            else float(soup.find(string="Maritime Link Import ").find_next("td").string)
+        )
+    except (AttributeError, TypeError):
+        logger.warning("unable to extract exchange data (error: {error})")
+        exchange = None
```

Comment:
I actually preferred the 'parser crashing' situation. Then you could easily see in the error log on which line the error occurred, and therefore which code needed updating without having to run and debug the parser first. (Now it can be any of the two, if I'm not mistaken.)

Perhaps this 'raise error from e' pattern I saw in another parser would work? If I'm not mistaken it might preserve the original stack trace and tell you on which line the error occurred.

https://github.com/electricitymaps/electricitymaps-contrib/blob/76b174e650103f6f9bfbf5f03810aa846022aaa7/parsers/ENTSOE.py#L1020-L1027

Reply:
Agreed, and yes it will perserve the original stack trace which helps with debugging 🙂

Reply:
Whoops, I had botched that anyway because I hadn't bound `error` (e.g., `except ... as error:`). I like this `raise ... from error` pattern; thanks for suggesting it!